Switch to tab if URL is open

jcmuller opened the pull request:
Switch to an already-open tab instead of opening duplicates

Add a `tabs.switch_to_open_url` setting, `off` by default. When it is `on`, opening a URL that is already loaded in some tab focuses that tab and its window instead of opening a second copy. The check runs after URL validation in `_open()`, in the `openurl()` current-tab path, and in `app.open_url()` for command-line and IPC opens.

A switch only considers tabs in the same privacy context, so opening a normal URL never jumps to a private-browsing tab. Taking a tab with `:tab-take` passes reuse=False so it is never turned into a switch.

The `:open` completion also shows a tabs category whenever tabs is in `completion.open_categories`.

Discussed on #2198 

*edit by @The-Compiler: Fixes #5037*

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/4955)
<!-- Reviewable:end -->

jgkamat commented on qutebrowser/completion/models/urlmodel.py:
```diff
@@ -86,6 +88,10 @@ def url(*, info):
             'Bookmarks', bookmarks, delete_func=_delete_bookmark, sort=False)
 
+    if buffers and 'buffers' in categories:
+        models['buffers'] = listcategory.ListCategory(
```
I haven't looked into it too much, but I think that some of the work in [miscmodels](http://www.github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/completion/models/miscmodels.py#L125) can be re-used here (avoiding the functions added to utils). You would also get the delete function for free as well.

However, you would need to find a solution to avoid breaking each window into its own category

jcmuller replied:
I've been trying to figure out how to reuse the `miscmodels` work, but the biggest problem I see right now is that the tuples are in different order. I've been toying with the idea of having a lambda transformer to get the tuples into the right shape for this function.

Thoughts?

jgkamat commented on qutebrowser/browser/commands.py:
```diff
@@ -303,6 +303,14 @@ def openurl(self, url=None, related=False,
                 tab = False
                 bg = True
 
+            if utils.url_in_tabs(cur_url):
```
I don't like this method of determining if we clicked on a "buffer" in the list. However, I don't think there's an easy way to fix it without redoing a lot of things (QStandardItem seems to only take a string).

Why not place this code inside `_open` behind a config switch? Then it would apply to `quickmark-load` and `bookmark-load` for free, it wouldn't cause any overhead for people disabling it, and it would be implied this behavior would be the case globally for all opened urls.

jcmuller replied:
I moved the check to `_open` and I did get the quickmark and bookmark as expected. But, when doing a plain `:open [existing url]` it did not kick in.

jgkamat commented on qutebrowser/browser/commands.py:
```diff
@@ -303,6 +303,14 @@ def openurl(self, url=None, related=False,
                 bg = True
 
+            if utils.url_in_tabs(cur_url):
+                index = utils.tuple_for_url_in_tab(cur_url.toDisplayString())[2]
```
This snippet is looping over all tabs 4 times

- generating the urls in `list_buffers`
- searching generated list in `url_in_tabs`
- generating the urls again in `list_buffers`
- equality check in `tuple_for_url_in_tab`.

If there's no way around avoiding the iteration completely, please just return the tab in the first loop, or if you must, generate a map from url->tab.

jcmuller replied:
This is now iterating only once.

jgkamat commented on qutebrowser/completion/models/urlmodel.py:
```diff
@@ -67,6 +68,7 @@ def url(*, info):
     quickmarks = [(url, name) for (name, url)
                   in objreg.get('quickmark-manager').marks.items()]
     bookmarks = objreg.get('bookmark-manager').marks.items()
+    buffers = utils.list_buffers()
```
If this sticks around, there's no reason to do this computation for everyone, we only need to do it if buffers is in completions.

jcmuller replied:
Done.

jgkamat commented on qutebrowser/config/configdata.yml:
```diff
@@ -958,13 +958,14 @@ downloads.location.suggestion:
     - quickmarks
     - bookmarks
     - history
+    - buffers
```
FWIW, I don't think that buffers should be here by default, and if it should, it should be at the top.

toofar commented:
I don't thing we should be increasing use of the "buffers" terminology. Yes I realiee I added the original buffer command but I feel like if someone looks at the config option and sees that buffer is an option they will not know what a buffer is. Tabs are tabs.

jcmuller commented:
I agree with you. I just wanted to be consistent. I will change to `Tabs`.

rcorre commented on qutebrowser/browser/commands.py:
```diff
@@ -1748,3 +1756,19 @@ def fullscreen(self, leave=False, enter=False):
         log.misc.debug('state before fullscreen: {}'.format(
             debug.qflags_key(Qt, window.state_before_fullscreen)))
+
+    def _possibly_open_in_existing_tab(self, url):
```
This is used in app.py, so don't mark it as private.

jcmuller replied:
Good catch

rcorre commented on qutebrowser/browser/commands.py:
```diff
@@ -1748,3 +1756,19 @@ def fullscreen(self, leave=False, enter=False):
+        if not config.val.tabs.switch_to_tab_on_url_match:
+            return False
+
+        tuple = utils.tuple_for_url_in_tab(url.toDisplayString())
```
`tuple` is a builtin type, don't shadow it.

rcorre commented on qutebrowser/browser/commands.py:
```diff
@@ -1748,3 +1756,19 @@ def fullscreen(self, leave=False, enter=False):
+        tuple = utils.tuple_for_url_in_tab(url.toDisplayString())
+
+        if tuple is not None:
+            index = tuple[2]
```
You call this `index`, but in `tuple_for_url_in_tab` is returning the `title` for the second field.

jcmuller replied:
It was indeed the index. But the function now just returns it, instead of the whole tuple.

rcorre commented on qutebrowser/utils/utils.py:
```diff
@@ -740,3 +741,29 @@ def ceil_log(number, base):
+    return tabs
+
+
+def tuple_for_url_in_tab(cur_url):
```
Unless I'm missing something, you only call this in one place, and that place only uses the `title`, so why not just return that?

jcmuller replied:
It only used the index (`0/1`). It now just returns it.

rcorre reviewed (changes requested):
Overall looks good! I think it is definitely worth trying to use a common util function for both `buffer` and `url` completion, and just have one reorder it. I _think_ you could return a `namedtuple` and still pass it directly to a `ListCategory`, which would improve readability for the one that has to reorder, e.g. 

```
tabs = [(tab.url, tab.title, tab.idx) for tab in util.list_tabs()]
```

jcmuller commented:
@rcorre thanks for pointing me towards `namedtuple`. It fits the bill handsomely.

I decided to introduce a helper class whose only responsibility is find tabs and can return different types of structures, depending on the need.

Is `utils` the best place for it? Now it is shared by the things that need to know about buffers.

jgkamat commented on qutebrowser/completion/models/urlmodel.py:
```diff
@@ -49,6 +50,13 @@ def _delete_quickmark(data):
     log.completion.debug('Deleting quickmark {}'.format(name))
     quickmark_manager.delete(name)
 
+def _delete_tab(data):
```
Is this being used?

jcmuller replied:
Removed.

jgkamat commented on qutebrowser/misc/tabsfinder.py:
```diff
@@ -0,0 +1,82 @@
+    def delete_tab(self, data):
+        """Close the selected tab."""
+
+        if self._tabbed_browser is None:
```
`self._tabbed_browser` is only really used here. I don't think it makes sense to pass it in the constructor to the class, but rather to this function directly (so it can't be called without specifying a tabbed browser).

Also, I don't think this function needs to strictly be a class method, it might be fine with just being a static method or not in this class at all, so we can call it without needing to run `_find`.

jcmuller replied:
Fixed.

rcorre commented on qutebrowser/config/configdata.yml:
```diff
@@ -1777,6 +1777,11 @@ tabs.wrap:
+tabs.switch_to_tab_on_url_match:
+  type: Bool
+  default: false
+  desc: Switch to tab in case the desired URL is already open
```
nit: all descriptions end with a `.`

rcorre commented on qutebrowser/completion/models/urlmodel.py:
```diff
@@ -86,6 +88,12 @@ def url(*, info):
         models['bookmarks'] = listcategory.ListCategory(
             'Bookmarks', bookmarks, delete_func=_delete_bookmark, sort=False)
 
+    if config.val.tabs.switch_to_tab_on_url_match and 'tabs' in categories:
```
I wonder if this should just be `'tabs' in categories`. I'm not sure if there's a good use case for completing tabs if `switch_to_tab_on_url_match` isn't set, but it could be confusing if setting `tabs` alone doesn't control this category.

If we keep it like this, at least document that _both_ need to be set for the category to appear.

jcmuller replied:
Good questions. I don't think I can answer this myself :)

jcmuller replied:
This has been addressed.

rcorre commented on qutebrowser/misc/tabsfinder.py:
```diff
@@ -0,0 +1,92 @@
+Tab = namedtuple('Tab', 'url title index')
+
+
+class TabsFinder:
```
Here's my 2 cents on this, though others may have a different philosophy :grin: 

I don't love the use of a class here, because:

1. It computes several different data structures, only one of which is necessary each place it is used.
2. It does non-trivial work in a constructor, which can be surprising
3. I'm not a fan of noun-ifying verbs. That is, don't use a class when a function would do (this probably qualifies as bikeshedding, though)

I would be more keen on one or more module-level functions:

1. `all_tabs`, which yields* successive (index, url, title) tuples.
2. `all_tabs_by_window` which groups `all_tabs` by `win_id` _or_ just omit this and do the grouping in the completion code

`index_for_url` probably  doesn't need a helper, it should be as simple as `idx = next((t for t in all_tabs()) if t.url == url), None)`

\* I think it makes sense as a generator since the evaluation will be transformed (for completion) or stopped early (for finding a matching tab). That being said, this isn't likely to be performance critical, so if using generators for some reason complicates things then ignore me :grin:

rcorre replied:
Maybe also consider returning (win_id, index, url, title) tuples, as that feels more "pure-data"-y, and only form strings when needed for display. Ignore this if it makes things more complicated though

rcorre commented on qutebrowser/misc/tabsfinder.py:
```diff
@@ -0,0 +1,92 @@
+
+    def _find(self):
+        """Find all the tabs
+
```
nit: extra whitespace.

rcorre commented on qutebrowser/misc/tabsfinder.py:
```diff
@@ -0,0 +1,92 @@
+
+    if type(data) is str:
+        index = data
+    elif type(data) is Tab:
```
This is only ever called from a completion category, which means the type should always be `list`, right?
Why the switches for other data types?

Functions that operate on a wide array of types are tricky, and I _think_ there's a movement towards adding typing to parts of qutebrowser now?

rcorre reviewed (changes requested):
More comments/opinions, but this is looking good!

jcmuller commented:
@rcorre I believe I addressed all your comments. Thanks for the feedback!

rcorre reviewed (changes requested):
Excellent! The implementation looks fine to me, now we just need tests.

- Add a test to `tests/unit/completion/test_models.py` to check `urlmodel` with `tabs`
- Create `tests/unit/misc/test_tabsfinder`
- Test the new behavior of `open`. I _think_ this is only tested at the end-to-end level now, at `tests/end2end/features/open.feature`

The-Compiler commented on qutebrowser/app.py:
```diff
@@ -333,6 +333,13 @@ def open_url(url, target=None, no_raise=False, via_ipc=True):
     log.init.debug("About to open URL: {}".format(url.toDisplayString()))
+
+    if config.val.tabs.switch_to_open_url:
+        cd = commands.CommandDispatcher(win_id, tabbed_browser)
```
It seems quite odd that `app.py` needs to create a `CommandDispatcher` for something which isn't actually a command being run (also, I'd like to get rid of `commands.py` entirely at some point: #1129).

Also see my comments below.

jcmuller replied:
Moved.

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -1748,3 +1756,19 @@ def fullscreen(self, leave=False, enter=False):
         log.misc.debug('state before fullscreen: {}'.format(
             debug.qflags_key(Qt, window.state_before_fullscreen)))
+
+    def possibly_open_in_existing_tab(self, url):
```
I see various issues with this function:

- It doesn't really belong here - it's not a command, and it's also used from outside `commands.py`. Why not move it to `tabsfinder.py`? Or maybe in `mainwindow.py` seeing that it already has various window utility functions? Maybe it'd even make sense to add a new `tabutils.py` instead of `tabsfinder.py`?
- Its name doesn't convey that it switches to a tab. It sounds like it'd check if an URL is possibly open in another tab, but it does more than that. Maybe a name like `switch_to_existing_tab` or so woudl be better?

jcmuller replied:
Moved and renamed.

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -1748,3 +1756,19 @@ def fullscreen(self, leave=False, enter=False):
+
+        if tab is not None:
+            index = "{}/{}".format(tab.win_id, tab.index)
+            self.buffer(index)
```
Having to construct a string index which is then parsed by the `:buffer` command again seems odd. Why not add a `switch_to_tab(win_id, tab_id)` or something to `tabsfinder.py` (or `tabutils.py`), then use that from `:buffer` and also use it from here?

jcmuller replied:
Done.

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -1748,3 +1756,19 @@ def fullscreen(self, leave=False, enter=False):
+            index = "{}/{}".format(tab.win_id, tab.index)
+            self.buffer(index)
+            message.info("URL is open already. Switching to it. {}".
+                         format(index))
```
This seems like an odd message to show ("URL is open already. Switching to it. 02/15"). Is it necessary to show it at all? If so, should the index be included?

jcmuller replied:
Removed.

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -21,7 +21,10 @@
+import itertools
 from qutebrowser.completion.models import completionmodel, listcategory, util
+from qutebrowser.misc import tabsfinder
+from qutebrowser.misc.tabsfinder import delete_tab
```
Please only import modules, not things form inside modules. With the import above, you can just use `tabsfinder.delete_tab` anyways.

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -100,30 +103,14 @@ def _buffer(skip_win_id=None):
-                         tab.url().toDisplayString(),
-                         tabbed_browser.widget.page_title(idx)))
+    for win_id, tabs in itertools.groupby(tabsfinder.all_tabs(skip_win_id), lambda
+            t: t.win_id):
```
nitpick: This would probably be more readable with the line break after the comma.

jcmuller replied:
Done

The-Compiler commented on qutebrowser/completion/models/urlmodel.py:
```diff
@@ -23,7 +23,8 @@
 from qutebrowser.config import config
-
+from qutebrowser.misc import tabsfinder
+from qutebrowser.misc.tabsfinder import delete_tab
```
Same as above.

jcmuller replied:
Fixed

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -21,7 +21,10 @@
 
 from qutebrowser.config import configdata
 from qutebrowser.utils import objreg, log
+import itertools
```
Please group the imports as per [pep8](https://www.python.org/dev/peps/pep-0008/#imports):

```
import itertools

from qutebrowser.… import …
```

The-Compiler commented on qutebrowser/misc/tabsfinder.py:
```diff
@@ -0,0 +1,43 @@
+
+from qutebrowser.utils import objreg
+
+Tab = namedtuple('Tab', 'win_id index url title')
```
Namedtuples are kind of confusing because they can also be used as a tuple even if they aren't.

Please use the [attrs](http://www.attrs.org/en/stable/) module instead (search for `@attr.s` in qutebrowser's codebase for some examples).

jcmuller replied:
Nice suggestion. Done.

The-Compiler commented on qutebrowser/misc/tabsfinder.py:
```diff
@@ -0,0 +1,43 @@
+    for win_id in objreg.window_registry:
+        tabbed_browser = objreg.get('tabbed-browser', scope='window', window=win_id)
+
+        if skip_win_id is not None and win_id == skip_win_id:
```
Since `win_id` will not be `None` anyways, you can drop the `if skip_win_id is not None` part.

The-Compiler commented on qutebrowser/misc/tabsfinder.py:
```diff
@@ -0,0 +1,43 @@
+            tab = tabbed_browser.widget.widget(idx)
+            t = Tab(int(win_id), idx + 1, tab.url(),
+                    tabbed_browser.widget.page_title(idx))
+            yield(t)
```
`yield t` for consistency with other `yield` (and `return`) statements.

The-Compiler commented on qutebrowser/misc/tabsfinder.py:
```diff
@@ -0,0 +1,43 @@
+
+        for idx in range(tabbed_browser.widget.count()):
+            tab = tabbed_browser.widget.widget(idx)
+            t = Tab(int(win_id), idx + 1, tab.url(),
```
Why the `int()`?

jcmuller replied:
Removed

The-Compiler commented on qutebrowser/misc/tabsfinder.py:
```diff
@@ -0,0 +1,43 @@
+Tab = namedtuple('Tab', 'win_id index url title')
+
+
+def all_tabs(skip_win_id=None):
```
Please add a docstring.

The-Compiler commented on qutebrowser/misc/tabsfinder.py:
```diff
@@ -0,0 +1,43 @@
+
+
+def delete_tab(data):
+    """Close the selected tab."""
```
This seems confusing to me (without having to look at lots of other code). What's the "selected" tab? What do I need to pass as "data" to this function? Maybe mention that it's intended to be used as `delete_func` for the completion?

jcmuller replied:
Changed.

The-Compiler commented on qutebrowser/misc/tabsfinder.py:
```diff
@@ -0,0 +1,43 @@
+    if re.match(pattern, data[0]):
+        index = data[0]
+    elif re.match(pattern, data[2]):
+        index = data[2]
```
This seems weird. Why use a regex to try and find something to close? Instead of having to guess, the caller should tell the function where to find the ID (for example by using [functools.partial](https://docs.python.org/3/library/functools.html#functools.partial) to get a function usable for `delete_func`).

The-Compiler commented on qutebrowser/misc/tabsfinder.py:
```diff
@@ -0,0 +1,43 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
```
This should also have the license comment other files have (with 2019 as the year).

The-Compiler reviewed (changes requested):
Thanks for your contribution! I added some comments about things I found odd while reading through the code - also please take a look at the [failing tests/linters](https://travis-ci.org/qutebrowser/qutebrowser/builds/570101889) and add tests like @rcorre mentioned.

jcmuller commented:
@rcorre @The-Compiler I think I've successfully introduced all your great feedback. Also, the builds are all green. Is there anything else you guys would suggest I change? Do you guys typically sqash PR commits?

The-Compiler commented:
I'll give this another review ASAP but it'll probably take a couple of weeks - I have my last exam coming up next week, then some holidays, and then a big PR backlog to tackle :wink:

The-Compiler commented:
From @toofar in https://github.com/qutebrowser/qutebrowser/pull/4807#discussion_r315094422

> I think that class would be more useful for general use if it returned actual AbstractTabs which is already a public API rather than creating a data class that needs to be added to every time a new use for it comes up.

I think that's a good point - do we really need another tab class when we already have one?

jcmuller commented:
Good point. I must admit I didn't know about `AbstractTabs`. I will use that class instead.

The-Compiler commented:
Note the object you get from `tabbed_browser.widget.widget(idx)` should already be an `AbstractTab` (either a `WebKitTab` or a `WebEngineTab`).

jcmuller commented:
Done.

rcorre commented:
> Do you guys typically sqash PR commits?

You don't have to. I sometimes collapse little fixup commits, but I think it is allowed (encouraged?) to keep PRs broken up into multiple commits where it is logical.

rcorre reviewed (commented):
A few more comments, none necessarily critical. It does seem there is [at least one](https://travis-ci.org/qutebrowser/qutebrowser/jobs/576347978#L3030) test failure. It seems `test_buffer_with_wrong_argument_1` is failing now when it tries to call `:buffer -1`. Instead of showing an error it seems to be crashing.

The-Compiler commented:
> > Do you guys typically sqash PR commits?
> 
> You don't have to. I sometimes collapse little fixup commits, but I think it is allowed (encouraged?) to keep PRs broken up into multiple commits where it is logical.

That's what I do (and prefer) myself as well. Little typo/linting fixes (where it's easy to do so) are best collapsed to the commit which introduced the issue, but if that's too hard to do, a dedicated "Fix lint" commit is fine as well.

For more substantial changes, I prefer having things split up, so it's possible to understand how something came into existence - and even go back to an earlier approach if needed for some reason. No point IMHO to destroy that information which can be quite useful.

I guess that's why qutebrowser has over 18k commits :laughing:

The-Compiler commented:
Looks like the [tests are failing](https://travis-ci.org/qutebrowser/qutebrowser/jobs/576347978) now - could you take a look please?

lucc commented:
I am sorry to jump into this discussion quite late. I proposed a different approach to the same problem some time ago: #3194  

My suggestion was to introduce a new command instead of a config option. I used my patch ever since and today I decided to open #5046 and only found this one afterwards.

There is one scenario I would like to know how well it is supported with this config option:
1. I have https://wikipedia.org/foobar open in some tab
2. now I want to open https://wikipedia.org/foo which is a substring of one of my open tabs but I want to open exactly this typed url.
3. what happens with the config option?

(With to different commands I can just use `:open` instead of `:goto` but if #836 is fixed I would prefer an option to `:open` over a separate command as it is even shorter and easier to change on the prompt.)

jgkamat commented:
Lucas Hoffmann writes:

> There is one scenario I would like to know how well it is supported with 
> this config option:
> 1. I have https://wikipedia.org/foobar open in some tab
> 2. now I want to open https://wikipedia.org/foo which is a substring of one 
> of my open tabs but I want to open exactly this typed url.
> 3. what happens with the config option?

If you don't have https://wikipedia.org/foo, it will open it as normal (and 
not switch).

> (With to different commands I can just use `:open` instead of `:goto` but if 
> #836 is fixed I would prefer an option to `:open` over a separate command as 
> it is even shorter and easier to change on the prompt.)

This solution works with all commands (such as quickmark open and bookmark 
open) which I would prefer. The other solution, since it adds a new command, 
would only work in that specific case.

lucc commented:
@jgkamat The point of my scenario was that I *do have and want to have* the other url open. But I found the `noreuse` option and have updated my comment above.

rcorre commented:
> This solution works with all commands 

It just occured to me that this solution probably also affects `tab-clone`, which might be one command we _don't_ want it to affect, since you're explicitly saying "I want a new copy of this tab".

jgkamat commented:
Ryan Roden-Corrent writes:

> It just occured to me that this solution probably also affects `tab-clone`, 
> which might be one command we _don't_ want it to affect, since you're 
> explicitly saying "I want a new copy of this tab".

tab-clone is unaffected here, as it does not use _open. If there is a command 
which (like tab-clone) wouldn't make sense with switching, we can just add the 
noreuse arg to the call.

jcmuller commented:
@The-Compiler @rcorre I finally had time to work on the failure I was getting. I think this is now ready. 

Cheers!

jcmuller commented:
The current failures are also failing on master.

The-Compiler commented:
Argh, looks I broke a couple of things while adding type annotations over the last couple of days. I'll look into it, sorry for that!

The-Compiler commented:
Should be fixed now - could you merge master into your branch (or rebase it)? Note that you'll need to add type annotations at least for `utils.tabutils` (optionally also for the rest of the code), as anything under `utils.*` now has type annotations (and that's enforced).

jcmuller commented:
Tanks! And thanks for the heads up. I’ll take care of it.

Cheers

On Sun, Oct 13, 2019 at 12:36 Florian Bruhin <notifications@github.com>
wrote:

> Should be fixed now - could you merge master into your branch (or rebase
> it)? Note that you'll need to add type annotations at least for
> utils.tabutils (optionally also for the rest of the code), as anything
> under utils.* now has type annotations (and that's enforced).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qutebrowser/qutebrowser/pull/4955?email_source=notifications&email_token=AABS45FBRCNIFDSYUWW3FCDQONFBNA5CNFSM4IJSIDMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBC2BKY#issuecomment-541434027>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABS45GMIZBMOMGAPSJC7ATQONFBNANCNFSM4IJSIDMA>
> .
>
-- 
Juan C.

jcmuller commented on qutebrowser/browser/commands.py:
```diff
@@ -120,8 +120,16 @@ def _open(self, url, tab=False, background=False, window=False,
             window: Whether to open in a new window
             private: If opening a new window, open it in private browsing mode.
                      If not given, inherit the current window's mode.
+            noreuse: If tabs.switch_to_open_url is enabled, do not switch to
```
Good call. Swapped.

jcmuller commented on qutebrowser/utils/tabutils.py:
```diff
@@ -0,0 +1,160 @@
+from qutebrowser.utils import objreg
+
+
+def all_tabs(skip_win_id=None):
```
Spent a little time checking this out. I couldn't get the return of `_buffer(skip_win_id)` to not include the skipped window id mapped to an empty list (`'0': [],`, for example).

I did like your idea and that would simplify tests.... but I'll leave it as is for now.

jcmuller commented on qutebrowser/utils/tabutils.py:
```diff
@@ -0,0 +1,160 @@
+
+def tab_for_url(url):
+    """Returns the tab that has the URL open."""
+    if not config.val.tabs.switch_to_open_url:
```
Good call. Done.

The-Compiler commented:
Looks like a lot of tests are failing on Travis/AppVeyor which pass on master - could you take a look at those please?

jcmuller committed:
tests: delete the mode manager synchronously in its fixture

The mode_manager fixture tore its ModeManager down with deleteLater(),
which never actually runs in a headless test (there is no event loop to
deliver it). So the manager survived teardown, and so did its key parsers'
QApplication-parented timers. Mostly harmless... until a timer-free test
runs right before test_timer and trips its leak guard.

sip.delete() drops it on the spot instead, the same way the tab-widget
fixture already does.

jcmuller committed:
Switch to an already-open tab instead of opening duplicates

Add a tabs.switch_to_open_url setting, off by default. When it is on,
opening a URL that is already loaded in some tab focuses that tab and
its window instead of opening a second copy. The check runs after URL
validation in _open(), in the openurl() current-tab path, and in
app.open_url() for command-line and IPC opens.

A switch only considers tabs in the same privacy context, so opening a
normal URL never jumps to a private-browsing tab. Taking a tab with
:tab-take passes reuse=False so it is never turned into a switch.

The :open completion also shows a tabs category whenever tabs is in
completion.open_categories.

kosciej16 commented:
In case you will try to merge it. I did it on my repository (ignored completion for now) and [found some issues](https://github.com/kosciej16/qutebrowser/commit/f875ed3da9f0ff413afc5a696f7be72a9f8ce42a) that had to be fixed. After that changes it's working for me (however I would expect less restricted condition, like depending on domain, not full url)

magnetophon commented:
Any news on this?
This is the feature I miss the most at the moment in this wonderful browser.

The-Compiler commented:
Not so far - I'll get back to this once I looked at more straightforward PRs first.

edi9999 commented:
I just found this issue and would really like to see this coming to qutebrowser @The-Compiler !

I just saw this for chromium : 
https://unix.stackexchange.com/questions/237626/is-there-a-way-to-activate-a-particular-tab-of-chrome-via-bash

I'm not sure though if after merging, this, it would also be possible to move to an existing tab from the command line (or create the tab if it does'nt exist).

edi9999 commented:
I just also found this question : https://listi.jpberlin.de/pipermail/qutebrowser/2016-January/000141.html

Since I wanted to try to achieve this using a user script.

Maybe one way would be to be able to get results from a command from a user script.

Something like this : 

```
echo "tab-list" >> "$QUTE_FIFO"
tab_list="$(cat "$QUTE_INPUT_FIFO")"
echo "$tab_list" # prints the same as what tab-select prints : 
# 1 https://github.com/qutebrowser/qutebrowser/pull/4955
# 2 https://start.duckduckgo.com/
```

The-Compiler commented:
@edi9999 see #902 on that.

edi9999 commented:
If someone else want do achieve this, here's how you can do it :+1: 

It works by first trying to run `qutebrowser ":tab-select $url"` and if that one fails, it will add to the log file which will then be read using the tail -f which then will start `qutebrowser "$url"` instead


```
qute_log=/tmp/qute.log
running=false
logfile="$qute_log"
if pgrep qutebrowser >/dev/null
then
	running=true
	logfile=/dev/null
fi

if [ "$running" = "false" ]
then
    qutebrowser "$url" >>"$logfile" 2>&1
    exit 0
fi
if ! [ -f "$qute_log" ]
then
    echo "" >>"$qute_log"
fi
last_log_time="$(tail -n 1 "$qute_log" | sed -E 's/^([0-9]:+).*/\1/g')"
qutebrowser ":tab-select $url" >>"$logfile" 2>&1
sleep 0.01
if tail -n 1 <"$qute_log" | tgrep -v "$last_log_time" | grep --fixed-strings "No matching tab for: $url"
then
    qutebrowser "$url" >"/dev/null" 2>&1
fi
```

jcmuller committed:
fix posargs attribute for pylint

jcmuller commented:
@The-Compiler @rcorre  sorry for the long time without reply. Life.... I was finally able to focus some time on cleaning this up and adapting it to the current state of the codebase. I've implemented all the suggestions and fixed all the fixable tests -- some remaining failures are completely not related from what I can see.

lucc commented on tests/end2end/features/open.feature:
```diff
@@ -133,3 +133,49 @@ Feature: Opening pages
         And I run :open quickmarktest
         Then data/numbers/10.txt should be loaded
+
+    Scenario: :open an open URL with tabs.switch_to_open_url set
```
Is there any difference between this test and the next one? I don't see any.

jcmuller replied:
line 146 is the difference.

this one opens a new tab. The next one opens in the current one.

jcmuller committed:
Consolidate the switch-to-open-URL check into a helper

The check (config flag, tab_for_url, switch_to_tab) lived in three spots:
app.open_url, commands._open, and the commands.openurl loop. They had
started to drift, only _open carried the reuse guard, so move the policy
into tabutils.switch_to_open_url and let each caller keep its own follow-up.

No behavior change.
